src/functional/fencil_processors/processor_interface.py

```diff
@@ -0,0 +1,155 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+"""
+Interface for fencil processors.
+
+Fencil processors are functions which take an ``iterator.ir.FencilDefinition`` along with input values for the fencil.
+Those which execute the fencil with the given arguments (possibly by generating code along the way) are fencil executors.
+Those which generate / any kind of string based on the fencil and (optionally) input values are fencil formatters.
+
+For more information refer to ``gt4py/docs/functional/architecture/007-Fencil-Processors.md``
+"""
+from __future__ import annotations
+
+from functools import update_wrapper
+from typing import Protocol, TypeVar, runtime_checkable
+
+from functional.iterator.ir import FencilDefinition
+
+from .source_modules import SourceModule
+
+
+PROCESSOR_RETURN_T = TypeVar("PROCESSOR_RETURN_T", covariant=True)
+PROCESSOR_KIND_T = TypeVar("PROCESSOR_KIND_T", bound="FencilProcessorProtocol", covariant=True)
+
+
+class FencilProcessorFunction(Protocol[PROCESSOR_RETURN_T]):
+    def __call__(self, fencil: FencilDefinition, *args, **kwargs) -> PROCESSOR_RETURN_T:
+        ...
+
+
+@runtime_checkable
```

Comment:
I don't think it makes a lot of sense to make this protocol runtime_checkable since it will only check if the object is callable and if it has a `kind` member, but the important check here is the actual value in `kind`. I propose to delete the annotation and instead create typeguards functions for each of the kinds which can be used at runtime and they also provide information to the type checker. For example, if `kind` is converted to a property then it can be as simple as:

```python
def is_fencil_formatter(obj: object, ) -> TypeGuard[FencilFormatter]:
    return callable(obj) and getattr(obj, "kind", None) is FencilFormatter

```

Reply:
I actually thought about changing it to a property. Seems to be the way to go.